### ISSUE TITLE
Support for other AbstractFloat types

### ISSUE THREAD

**Issue by giordano** [kbarbary/SkyCoords.jl#3]
Is there any plan to support AbstractFloat types for sky coordinates?  I'd like to use this package in [`AstroLib`](https://github.com/JuliaAstro/AstroLib.jl), which doesn't prefer any floating point type, instead `SkyCoords` uses `Float64` everywhere.

**Comment by kbarbary** [comment]
Sounds like a good idea. Not sure I'll have the time for it, but happy to accept a PR.

**Comment by giordano** [comment]
I'm playing with this change, but it's a bit tricky, because if one wants to use directly the constructor has to manually specify the type.

How about changing the user interface (in a backward incompatible way, though)?  We could do something like this:

```julia
immutable ICRSCoords{T<:AbstractFloat} <: AbstractSkyCoords
    ra::T
    dec::T
end
icrscoords{T<:AbstractFloat}(ra::T, dec::T) = ICRSCoords(mod(ra, 2pi), dec)
icrscoords(ra::Real, dec::Real) = icrscoords(promote(float(ra), float(dec))...)

immutable GalCoords{T<:AbstractFloat} <: AbstractSkyCoords
    l::T
    b::T
end
galcoords{T<:AbstractFloat}(l::T, b::T) = GalCoords(mod(l, 2pi), b)
galcoords(l::Real, b::Real) = galcoords(promote(float(l), float(b))...)

# FK5 is parameterized by equinox (e)
immutable FK5Coords{e,T<:AbstractFloat} <: AbstractSkyCoords
    ra::T
    dec::T
end
fk5coords{T<:AbstractFloat}(ra::T, dec::T, eq::Real) =
    FK5Coords{eq,T}(mod(ra, 2pi), dec)
fk5coords(ra::Real, dec::Real, eq::Real=2000) =
    fk5coords(promote(float(ra), float(dec))..., eq)
```

You can suggest better names for the user-exposed functions.

**Comment by kbarbary** [comment]
Did you try using an outer constructor? Constructors for parametric types get a little confusing, but I think the follow should work:

```julia
immutable ICRSCoords{T<:AbstractFloat} <: AbstractSkyCoords
    ra::T
    dec::T
    ICRSCoords(ra, dec) = new(mod(ra, 2pi), dec)
end

ICRSCoords{T<:AbstractFloat}(ra::T, dec::T) = ICRSCoords{T}(ra, dec)
ICRSCoords(ra::Real, dec::Real) = ICRSCoords(promote(float(ra), float(dec))...)
```

with this I get:
```julia
julia> ICRSCoords(0, 0)
ICRSCoords{Float64}(0.0,0.0)

julia> ICRSCoords(0, 0.)
ICRSCoords{Float64}(0.0,0.0)

julia> ICRSCoords(0.0f0, 0.0f0)
ICRSCoords{Float32}(0.0f0,0.0f0)

julia> ICRSCoords(0.0f0, 0.0)
ICRSCoords{Float64}(0.0,0.0)
```

Seems to work as intended. The key insight is that when you override the default inner constructor (as we have), the default outer constructor is no longer created, so you need to manually define the `ICRSCoords{T<:AbstractFloat}(ra::T, dec::T)` outer constructor.

**Comment by kbarbary** [comment]
PS: In FK5Coords, lets put the `T` parameter before the `e`. That way it mimics `Array{T,N}`, perhaps the most familiar 2-parameter type.

**Comment by giordano** [comment]
Yeah, I did try but in a different way that didn't work.  Please, have a look at my pull request.  This can be probably improved, but it is a working starting point.

**Comment by kbarbary** [comment]
Closed by #4
